edgekv/src/schema.rs

```diff
@@ -0,0 +1,250 @@
+use chrono::Utc;
+use crc::{Crc, CRC_32_CKSUM};
+use std::io::Read;
+pub const CRC_CKSUM: Crc<u32> = Crc::<u32>::new(&CRC_32_CKSUM);
+use crate::Result;
+#[derive(Debug, Clone, PartialOrd, PartialEq)]
+pub struct DataEntry {
+    crc: u32,
+    timestamp: i64,
+    key_size: u64,
+    value_size: u64,
+    key: Vec<u8>,
+    value: Vec<u8>,
+}
+
+pub trait Encoder {
+    fn encode(&self) -> Vec<u8>;
+}
+
+pub trait Decoder {
+    fn decode<R: Read>(rdr: &mut R) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+impl Encoder for DataEntry {
+    fn encode(&self) -> Vec<u8> {
+        let content = self.encode_content();
+        let crc = CRC_CKSUM.checksum(&content);
+        let mut buf = vec![];
+        buf.extend_from_slice(&crc.to_be_bytes());
+        buf.extend_from_slice(&content);
+        return buf;
+    }
+}
+
+impl Decoder for DataEntry {
+    fn decode<R: Read>(rdr: &mut R) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let mut out = Self {
+            crc: 0,
+            timestamp: 0,
+            key_size: 0,
+            value_size: 0,
+            key: vec![],
+            value: vec![],
+        };
+        let mut raw_crc_bytes = [0_u8; 4];
+        let mut raw_timestamp_bytes = [0_u8; 8];
+        let mut raw_key_size_bytes = [0_u8; 8];
+        let mut raw_value_size_bytes = [0_u8; 8];
+
+        rdr.read_exact(&mut raw_crc_bytes)?;
+        rdr.read_exact(&mut raw_timestamp_bytes)?;
+        rdr.read_exact(&mut raw_key_size_bytes)?;
+        rdr.read_exact(&mut raw_value_size_bytes)?;
+
+        out.crc = u32::from_be_bytes(raw_crc_bytes);
+        out.timestamp = i64::from_be_bytes(raw_timestamp_bytes);
+        out.key_size = u64::from_be_bytes(raw_key_size_bytes);
+        out.value_size = u64::from_be_bytes(raw_value_size_bytes);
+
+        let mut raw_key_bytes = vec![0_u8; out.key_size as usize];
+        let mut raw_value_bytes = vec![0_u8; out.value_size as usize];
+
+        rdr.read_exact(&mut raw_key_bytes);
```

Comment:
These two here can result in an error but it is not being handled (but the other calls above are), why?